Implement equivalence cache by caching and re-using predicate result

resouer opened the pull request:
The last part of #30844, I opened a new PR instead of overwrite the old one because we changed some basic assumption by allowing invalidating equivalence cache item by individual predicate.

The idea of this PR is based on discussion in https://github.com/kubernetes/kubernetes/issues/32024

- [x]  Pods belong to same controllerRef considered to be equivalent
- [x] ` podFitsOnNode` will use cached predicate result if it's available
- [x] Equivalence cache will be updated when if a fresh new predicate is done
- [x] `factory.go` will invalid specific predicate cache(s) based on the object change
- [x] Since `schedule` and `bind` are async, we need to optimistically invalid affected cache(s) before `bind`
- [x] Fully unit test of affected files
- [x] e2e test to verify cache update/invalid workflow
- [x] performance test results

- [x] Some nits fixes related but expected to result in `needs-rebase` so they are split to: #36060 #35968 #37512

cc @wojtek-t @davidopp

<!-- Reviewable:start -->

---
This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/36238)
<!-- Reviewable:end -->

resouer commented:
@k8s-bot cvm gce e2e test this #IGNORE

resouer commented:
@k8s-bot kops aws e2e test this #IGNORE

resouer commented:
@k8s-bot cvm gce e2e test this #IGNORE

resouer commented:
Emm, can't make this pass e2e anyhow because always timeout. 

@wojtek-t Do you have any tips on this? I don't think it's newly added e2e caused this timeout (seems it hasn't been executed yet ...) The code is generally ready and I've been e2e tested it on GCE and it works fine.

wojtek-t commented:
@resouer - I'm afraid that if it is consistently failing (and a bunch of different suites on the same thing), this is really related to your PR.

resouer commented:
@wojtek-t  Emm, so I assume there's something wrong on scheduler. But there's no available log of components on Jenkins, and test those failed ones locally on my GCE does not fail. Let me run the whole e2e again locally, may be some inter influences occurred ...

resouer commented:
@k8s-bot unit test this #IGNORE

resouer commented:
@wojtek-t yeah, CI passed, rebase & a wrong type fix did the magic. 

PTAL if you have time. The code is general ready for a primitive review.

wojtek-t commented:
@resouer - cool; will do the review, but after the 1.5 release is out (so most probably will do the review early next week)

resouer commented:
@k8s-bot gci gke e2e test this #IGNORE

resouer commented:
@k8s-bot cri node e2e test this #IGNORE

resouer commented:
@k8s-bot unit test this

resouer commented:
@k8s-bot gce etcd3 e2e test this

smarterclayton commented:
Any estimates on the total difference this makes in either scheduler throughput or latency?

resouer commented:
@smarterclayton Benchmark shows two facts:
1. equivalence cache does not show too much difference when predicates only cost small latency:
```
(μs)
default scheduler without eclass:					
pod-affinity	346	338	464	349	346
rc-affnity	348	380	378	378	376
scheduler with ecalss:					
pod-affinity	394	334	390	339	353
rc-affnity	565	428	428	526	534
```
2. equivalence cache will save huge time when predicates cost significant latenct (affinity predicate has 50ms latency)
```
(μs)
default scheduler without eclass:
pod-affinity	50474	50474	50474	50474	50474
rc-affinity	53241	50766	50982	50944	50968
scheduler with ecalss:						
pod-affinity	50766	50766	50766	50766	50766
rc-affinity	50860	442	429	434	508
```

kargakis commented on plugin/pkg/scheduler/equivalence_cache.go:
```diff
@@ -61,74 +70,158 @@ func NewEquivalenceCache(getEquivalencePodFunc algorithm.GetEquivalencePodFunc)
-	hashutil.DeepHashObject(hash, equivalencePod)
-	return uint64(hash.Sum32())
+	if equivalencePod != nil {
+		hash := adler32.New()
```
I would suggest switching to fnv if possible.

resouer replied:
Is it because fnv is more effective?

kargakis replied:
fnv is slightly less effective but much more stable.
https://github.com/kubernetes/kubernetes/pull/39527
https://github.com/kubernetes/kubernetes/issues/29735#issuecomment-261961559

kargakis replied:
Opened https://github.com/kubernetes/kubernetes/issues/40140 since it turns out there are a couple of more places that adler is used.

kargakis replied:
Opened #40140 since it turns out there are a couple of more places where adler is used.

kargakis commented on plugin/pkg/scheduler/equivalence_cache.go:
```diff
@@ -61,74 +70,158 @@ func NewEquivalenceCache(getEquivalencePodFunc algorithm.GetEquivalencePodFunc)
+	ec.InvalidCachedPredicateItem(nodeName, invalidPredicates)
+
+	// MatchInterPodAffinity, this case is complicated, we'll invalid equivalence cache of all nodes
+	affinity, err := api.GetAffinityFromPodAnnotations(pod.Annotations)
```
These fields are moved in the pod spec, right?

resouer replied:
Yes, it has been moved to PodSpec, will update this part.

resouer commented:
@k8s-bot unit test this

resouer commented:
@k8s-bot unit test this

resouer commented:
@k8s-bot cvm gke e2e test this

resouer commented:
@k8s-bot gci gke e2e test this

resouer commented:
@k8s-bot cvm gke e2e test this

resouer commented:
@k8s-bot gci gke e2e test this

resouer commented on plugin/pkg/scheduler/factory/factory.go:
```diff
@@ -164,22 +165,79 @@ func NewConfigFactory(client clientset.Interface, schedulerName string, hardPodA
 	)
 
+	// TODO(harryz) Delete this part, existing equivalence cache should not be affected by add/delete RC/Deployment etc, it only make sense when
+	// pod is scheduled or deleted
```
I think this also applies to StatefulSet, so I will rebase #39718 in.

resouer commented:
@k8s-bot cvm gce e2e test this

resouer commented:
NOTE: statefulset test in `Jenkins kops AWS e2e` failed after enable eclass for SS.

resouer commented:
@jayunit100 I just updated equivalence cache to enable statefulset & fixed CI. Mind to take a look?

kargakis commented on plugin/pkg/scheduler/factory/factory.go:
```diff
@@ -164,22 +165,79 @@ func NewConfigFactory(client clientset.Interface, schedulerName string, hardPodA
+		},
+	)
+
+	c.pVCLister.Indexer, c.pvcPopulator = cache.NewIndexerInformer(
```
Can we use shared informers for this one?  Cc @ncdc

resouer replied:
Not quite clear what's the potential issue of shared informers in this case?

ncdc replied:
@kargakis yes, although the scheduler is the last one I'm planning on tackling because it has more challenges than the basic controllers.

@resouer we want to use shared informers any time we have 2 or more distinct places in the code that would both be creating and using an informer, *and* the code does not need to manually manipulate the informer's indexer/store.

kargakis replied:
@ncdc naive question: is the shared cache going to work if any of the master components is scheduled in a different node?

ncdc replied:
@kargakis it's per process

ncdc replied:
So the kube-controller-manager is the biggest win in terms of reuse. But other things can benefit too assuming they're in the same process and need to list/watch the same resource type.

Commit:
Update factory.go informers to update equivalence cache

Fix tombstone

Add e2e to verify equivalence cache

Addressing nits in factory,go and e2e

Update build files

resouer commented:
@kargakis @wojtek-t @jayunit100 I just split this path into 5 independent commits. I am also thinking  send the first commit (at least) as a separate PR, because it only affects `equivalence_cache.go` and will not break anything.

smarterclayton commented:
As a rule, use informer.  If using informer, use shared informers.  If your
use case doesn't look like an informer, look harder.  If at the very end of
that, and it still doesn't look like an informer, maybe use something else
after talking to someone.  But probably use informer.

On Feb 15, 2017, at 10:03 AM, Andy Goldstein <notifications@github.com>
wrote:

*@ncdc* commented on this pull request.
------------------------------

In plugin/pkg/scheduler/factory/factory.go
<https://github.com/kubernetes/kubernetes/pull/36238>:

> +			},
+			DeleteFunc: func(obj interface{}) {
+				invalidPredicates := sets.NewString("MaxPDVolumeCountPredicate")
+				pv := obj.(*v1.PersistentVolume)
+				if pv.Spec.AWSElasticBlockStore != nil {
+					invalidPredicates.Insert("MaxEBSVolumeCount")
+				}
+				if pv.Spec.GCEPersistentDisk != nil {
+					invalidPredicates.Insert("MaxGCEPDVolumeCount")
+				}
+				c.equivalencePodCache.InvalidCachedPredicateItemOfAllNodes(invalidPredicates)
+			},
+		},
+	)
+
+	c.pVCLister.Indexer, c.pvcPopulator = cache.NewIndexerInformer(

So the kube-controller-manager is the biggest win in terms of reuse. But
other things can benefit too assuming they're in the same process and need
to list/watch the same resource type.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/36238>, or mute the thread
<https://github.com/notifications/unsubscribe-auth/ABG_pzvakkduRMcYME59iJK1zZ6hPPtpks5rcxPVgaJpZM4Kpp4B>
.

ncdc commented:
> As a rule, use informer.  If using informer, use shared informers.  If your
use case doesn't look like an informer, look harder.  If at the very end of
that, and it still doesn't look like an informer, maybe use something else
after talking to someone.  But probably use informer.

+1 to that

resouer commented:
@k8s-bot gci gke e2e test this

resouer commented:
I will get #41541 in first, and then move back here to refactor this part to shared informer.

resouer commented:
#41541 Is in, so I will start from rebase and update factories to latest informer changes

wojtek-t commented:
@resouer - parts of this PR have already been merged.

Should we close this one completely? If not, can you please rebase it?

resouer commented:
@wojtek-t I've already finished rebase locally, and now I am testing, will update the commit very soon!

resouer commented:
@k8s-bot kops aws e2e test this

resouer commented:
@k8s-bot kops aws e2e test this

resouer commented:
@k8s-bot gce etcd3 e2e test this

resouer commented:
Just can not pass kops aws e2e, but did not find any clue ...

In scheduler, it reports the pod which failed the test has been scheduled successfully:
```
I0413 14:53:46.876262       5 event.go:217] Event(v1.ObjectReference{Kind:"Pod", Namespace:"e2e-tests-statefulset-lc2ms", Name:"ss-1", UID:"ff2a8448-2058-11e7-8227-02e5737a682a", APIVersion:"v1", ResourceVersion:"1691", FieldPath:""}): type: 'Normal' reason: 'Scheduled' Successfully assigned ss-1 to ip-172-20-57-62.us-west-1.compute.internal
```

But in kubelet, it reports volume is not in bound phase:
```
I0413 14:53:46.894784    2530 kubelet.go:1814] SyncLoop (ADD, "api"): "ss-1_e2e-tests-statefulset-lc2ms(ff2a8448-2058-11e7-8227-02e5737a682a)"
I0413 14:53:46.895210    2530 qos_container_manager_linux.go:285] [ContainerManager]: Updated QoS cgroup configuration
E0413 14:53:46.994461    2530 desired_state_of_world_populator.go:259] Error processing volume "datadir" for pod "ss-1_e2e-tests-statefulset-lc2ms(ff2a8448-2058-11e7-8227-02e5737a682a)": error processing PVC "e2e-tests-statefulset-lc2ms"/"datadir-ss-1": PVC e2e-tests-statefulset-lc2ms/datadir-ss-1 has non-bound phase ("Pending") or empty pvc.Spec.VolumeName ("")
```
I did not find any clue why this PV is not available on this node, any special case of this node so it should not run ss-1 but scheduler (eclass) did not aware?

kargakis commented on plugin/pkg/scheduler/factory/factory.go:
```diff
@@ -161,7 +163,82 @@ func NewConfigFactory(
+			},
+			DeleteFunc: func(obj interface{}) {
+				invalidPredicates := sets.NewString("MaxPDVolumeCountPredicate")
+				pv := obj.(*v1.PersistentVolume)
```
This can also be a tombstone, no?

resouer replied:
@kargakis Sorry, not quite get what tombstone means ... 😆

kargakis replied:
This can be either a PersistentVolume or a cache.DeletedFinalStateUnknown object. See `deletePodFromCache` for example.

resouer replied:
Oh, yes it is.

resouer commented:
@kargakis  Updated. Thanks!

Btw, any idea of the kops aws e2e failure? Still no clue about how to fix it ...

resouer commented:
@wojtek-t @jayunit100 @kargakis Just rebased and fixed the CI, please review :)

resouer commented:
Any time to back to this PR: @wojtek-t @jayunit100 @kargakis

wojtek-t commented on plugin/pkg/scheduler/factory/factory.go:
```diff
@@ -201,11 +203,118 @@ func NewConfigFactory(
+	)
+	c.pVCLister = pvcInformer.Lister()
+
+	// This is for ServiceAffinity: it will not be affected by svc add/delete, except that the selector of the service is updated.
```
nit: can you please split this line into two?

wojtek-t commented on plugin/pkg/scheduler/factory/factory.go:
```diff
@@ -201,11 +203,118 @@ func NewConfigFactory(
+	pvInformer.Informer().AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
+			// MaxPDVolumeCountPredicate: since it relies on the counts of PV.
+			AddFunc:    c.onPvAdd,
```
Why we don't register UpdateFunc?

Same for pvc.

resouer replied:
predicates related to PV/PVC are all about "counts", so I assume update is not related to this.

wojtek-t commented on plugin/pkg/scheduler/factory/factory.go:
```diff
@@ -240,6 +348,7 @@ func (c *ConfigFactory) addPodToCache(obj interface{}) {
 	if err := c.schedulerCache.AddPod(pod); err != nil {
 		glog.Errorf("scheduler cache AddPod failed: %v", err)
 	}
+	// NOTE: Updating equivalence cache of addPodToCache has been handled optimistically in InvalidateCachedPredicateItemForPodAdd.
```
What if the pod was e.g. manually scheduled by the user?

resouer replied:
If a pod can bypasse scheduler, then scheduling algorithms (including eclass) shouldn't have affected this pod e.g. static pod, am I right here?

wojtek-t commented on plugin/pkg/scheduler/factory/factory.go:
```diff
@@ -257,6 +366,21 @@ func (c *ConfigFactory) updatePodInCache(oldObj, newObj interface{}) {
 		glog.Errorf("scheduler cache UpdatePod failed: %v", err)
 	}
+
+	// if the pod does not have binded node, updating equivalence cache is meaningless;
```
Can you please split this into a separate function and just call it here?
e.g. invalidateCachedPredicatesOnUpdatePod or sth like that?

wojtek-t commented on plugin/pkg/scheduler/factory/factory.go:
```diff
@@ -278,6 +402,21 @@ func (c *ConfigFactory) deletePodFromCache(obj interface{}) {
 		glog.Errorf("scheduler cache RemovePod failed: %v", err)
 	}
+
+	// part of this case is the same as pod add.
```
Can you please split this into a separate function and just call it here?

wojtek-t commented on plugin/pkg/scheduler/factory/factory.go:
```diff
@@ -307,6 +448,54 @@ func (c *ConfigFactory) updateNodeInCache(oldObj, newObj interface{}) {
 		glog.Errorf("scheduler cache UpdateNode failed: %v", err)
 	}
+
+	// Begin to update equivalence cache based on node update
```
Please split it to a separate function.

wojtek-t commented on test/e2e/scheduling/equivalence_cache_predicates.go:
```diff
@@ -0,0 +1,287 @@
+/*
+Copyright 2016 The Kubernetes Authors.
```
2017

wojtek-t commented on test/e2e/scheduling/equivalence_cache_predicates.go:
```diff
@@ -0,0 +1,287 @@
+		framework.WaitForAllNodesHealthy(cs, time.Minute)
+		masterNodes, nodeList = framework.GetMasterAndWorkerNodesOrDie(cs)
+
+		err := framework.CheckTestingNSDeletedExcept(cs, ns)
```
framework.ExpectNoError(framework.CheckTesting...)

wojtek-t commented on test/e2e/scheduling/equivalence_cache_predicates.go:
```diff
@@ -0,0 +1,287 @@
+		// It is so because we need to have precise control on what's running in the cluster.
+		systemPods, err := framework.GetPodsInNamespace(cs, ns, ignoreLabels)
+		Expect(err).NotTo(HaveOccurred())
+		systemPodsNo = 0
```
This is our namespace just for this tests - there shouldn't be any pods in it. I don't really follow this. Can you explain?

wojtek-t replied:
OK, now I see that this was copied from SchedulerPredicates tests.
@gmarek - can you explain this?

wojtek-t commented on test/e2e/scheduling/equivalence_cache_predicates.go:
```diff
@@ -0,0 +1,287 @@
+
+	})
+
+	// This test verifies that GenericPredicates works as expected:
```
s/Generic/General/

wojtek-t commented on test/e2e/scheduling/equivalence_cache_predicates.go:
```diff
@@ -0,0 +1,287 @@
+	})
+
+	// This test verifies that GenericPredicates works as expected:
+	// When a replica pod (with HostPorts) is scheduled to a node, it will invalidate GenericPredicates cache on this node,
```
s/Generic/General/

wojtek-t commented on test/e2e/scheduling/equivalence_cache_predicates.go:
```diff
@@ -0,0 +1,287 @@
+	// When a replica pod (with HostPorts) is scheduled to a node, it will invalidate GenericPredicates cache on this node,
+	// so that subsequent replica pods with same host port claim will be rejected.
+	// We enforce all replica pods bind to the same node so there will always be conflicts.
+	It("validates GenericPredicates is properly invalidated when a pod is scheduled [Slow]", func() {
```
s/Generic/General/

wojtek-t commented on test/e2e/scheduling/equivalence_cache_predicates.go:
```diff
@@ -0,0 +1,287 @@
+	})
+
+	// This test verifies that MatchInterPodAffinity works as expected.
+	// In equivalence cache, it dose not handle inter pod affinity (anti-affinity) specially (unless node label changed),
```
s/dose/does/

wojtek-t commented on plugin/pkg/scheduler/factory/factory.go:
```diff
@@ -409,11 +600,17 @@ func (f *ConfigFactory) CreateFromKeys(predicateKeys, priorityKeys sets.String,
-	// TODO(resouer) use equivalence cache instead of nil here when #36238 get merged
-	algo := core.NewGenericScheduler(f.schedulerCache, nil, predicateFuncs, predicateMetaProducer, priorityConfigs, priorityMetaProducer, extenders)
+	// Init equivalence class cache
+	if getEquivalencePodFunc != nil {
```
@davidopp - do we want to enable it by default in 1.7? Or should we maybe flaggate it?

resouer replied:
I think maybe a feature gate is better.

wojtek-t replied:
Feature gate sounds good to me.

wojtek-t replied:
I would like to avoid enabling it by default now (especially if this will hit 1.7 release).

resouer commented:
@wojtek-t Thanks for review. I just addressed the latest comments.

wojtek-t commented on plugin/pkg/scheduler/factory/factory.go:
```diff
@@ -201,11 +203,119 @@ func NewConfigFactory(
+	c.pVCLister = pvcInformer.Lister()
+
+	// This is for ServiceAffinity: it will not be affected by svc add/delete,
+	// except that the selector of the service is updated.
```
Hmm, I'm not sure. What if we create a new service and suddenly a pod that wasn't part of any service becomes a member of a service? Same for delete.

resouer replied:
That's true.This seems to be a rescheduler case but anyway we need to invalidate the ServiceAffinity cache.

wojtek-t commented on plugin/pkg/scheduler/factory/factory.go:
```diff
@@ -201,11 +203,119 @@ func NewConfigFactory(
+	)
+	c.serviceLister = serviceInformer.Lister()
+
+	// NOTE(resouer) Existing equivalence cache should not be affected by add/delete RC/Deployment etc, it only make sense when
```
Remove "Note(resouer)".
Also, please split the line so that it's not that long.

wojtek-t commented on plugin/pkg/scheduler/factory/factory.go:
```diff
@@ -201,11 +203,119 @@ func NewConfigFactory(
 }
 
+func (c *ConfigFactory) onPvAdd(obj interface{}) {
+	c.invalidatePredicatesForPv(obj.(*v1.PersistentVolume))
```
If here is a bug somewhere and we pass an object different than v1.PV, then it will panic.
I would prefer checking if it can be cast to PV and if not log an error.

wojtek-t commented on plugin/pkg/scheduler/factory/factory.go:
```diff
@@ -201,11 +203,119 @@ func NewConfigFactory(
+}
+
+func (c *ConfigFactory) onPvcAdd(obj interface{}) {
+	c.invalidatePredicatesForPvc(obj.(*v1.PersistentVolumeClaim))
```
Same here about casting.

wojtek-t commented on plugin/pkg/scheduler/factory/factory.go:
```diff
@@ -201,11 +203,119 @@ func NewConfigFactory(
+
+func (c *ConfigFactory) invalidatePredicatesForPvc(pvc *v1.PersistentVolumeClaim) {
+	if pvc.Spec.VolumeName != "" {
+		c.equivalencePodCache.InvalidateCachedPredicateItemOfAllNodes(sets.NewString("MaxPDVolumeCountPredicate"))
```
Instead of creating the new "sets.NewString" every time, can we just make this a constant?

resouer replied:
I'm not very sure how to do this ... since we need to insert value into this set.

wojtek-t replied:
Can you do the same you did with "serviceAffinitySet" ?

wojtek-t commented on plugin/pkg/scheduler/factory/factory.go:
```diff
@@ -201,11 +203,119 @@ func NewConfigFactory(
+}
+
+func (c *ConfigFactory) onServiceUpdate(oldObj interface{}, newObj interface{}) {
+	// TODO(resouer) We may need to invalid this for specified pod only?
```
s/invalid/invalidate/

wojtek-t commented on plugin/pkg/scheduler/factory/factory.go:
```diff
@@ -201,11 +203,119 @@ func NewConfigFactory(
+	oldService := oldObj.(*v1.Service)
+	newService := newObj.(*v1.Service)
+	if !reflect.DeepEqual(oldService.Spec.Selector, newService.Spec.Selector) {
+		c.equivalencePodCache.InvalidateCachedPredicateItemOfAllNodes(sets.NewString("ServiceAffinity"))
```
Instead of creating the new "sets.NewString" every time, can we just make this a constant?

wojtek-t commented on plugin/pkg/scheduler/factory/factory.go:
```diff
@@ -257,6 +368,25 @@ func (c *ConfigFactory) updatePodInCache(oldObj, newObj interface{}) {
+	if len(newPod.Spec.NodeName) != 0 && newPod.Spec.NodeName == oldPod.Spec.NodeName {
+		if !reflect.DeepEqual(oldPod.GetLabels(), newPod.GetLabels()) {
+			// MatchInterPodAffinity need to be reconsidered for this node, as well as all nodes in its same failure domain.
+			c.equivalencePodCache.InvalidateCachedPredicateItemOfAllNodes(sets.NewString("MatchInterPodAffinity"))
```
Instead of creating the new "sets.NewString" every time, can we just make this a constant?

wojtek-t commented on plugin/pkg/scheduler/factory/factory.go:
```diff
@@ -257,6 +368,25 @@ func (c *ConfigFactory) updatePodInCache(oldObj, newObj interface{}) {
+	// if pod's binded node has been changed, that case should be handled by pod add & delete.
+	if len(newPod.Spec.NodeName) != 0 && newPod.Spec.NodeName == oldPod.Spec.NodeName {
+		if !reflect.DeepEqual(oldPod.GetLabels(), newPod.GetLabels()) {
+			// MatchInterPodAffinity need to be reconsidered for this node, as well as all nodes in its same failure domain.
```
Can you please split the line to make it shorter?

wojtek-t commented on plugin/pkg/scheduler/factory/factory.go:
```diff
@@ -257,6 +368,25 @@ func (c *ConfigFactory) updatePodInCache(oldObj, newObj interface{}) {
+		invalidPredicates := sets.NewString()
+		// if requested container resource changed, invalidate GeneralPredicates of this node
+		if !reflect.DeepEqual(predicates.GetResourceRequest(newPod), predicates.GetResourceRequest(oldPod)) {
+			invalidPredicates.Insert("GeneralPredicates")
```
Why not invalidate it here instead of creating this "invalidPredicates" variable?

wojtek-t commented on plugin/pkg/scheduler/factory/factory.go:
```diff
@@ -278,6 +408,25 @@ func (c *ConfigFactory) deletePodFromCache(obj interface{}) {
+	// if this pod have these PV, cached result of disk conflict will become invalid.
+	for _, volume := range pod.Spec.Volumes {
+		if volume.GCEPersistentDisk != nil || volume.AWSElasticBlockStore != nil || volume.RBD != nil || volume.ISCSI != nil {
+			invalidPredicates.Insert("NoDiskConflict")
```
Why not invalidate it here instead of creating this "invalidPredicates" variable?

wojtek-t commented on plugin/pkg/scheduler/factory/factory.go:
```diff
@@ -278,6 +408,25 @@ func (c *ConfigFactory) deletePodFromCache(obj interface{}) {
+func (c *ConfigFactory) invalidateCachedPredicatesOnDeletePod(pod *v1.Pod) {
+	// part of this case is the same as pod add.
+	c.equivalencePodCache.InvalidateCachedPredicateItemForPodAdd(pod, pod.Spec.NodeName)
+	// MatchInterPodAffinity need to be reconsidered for this node, as well as all nodes in its same failure domain.
```
Can you please split the line to make it shorter?

wojtek-t commented on plugin/pkg/scheduler/factory/factory.go:
```diff
@@ -307,6 +458,58 @@ func (c *ConfigFactory) updateNodeInCache(oldObj, newObj interface{}) {
+	// Begin to update equivalence cache based on node update
+	invalidPredicates := sets.NewString()
+
+	oldTaints, err := helper.GetTaintsFromNodeAnnotations(oldNode.GetAnnotations())
```
Handle this err too.

wojtek-t commented on plugin/pkg/scheduler/factory/factory.go:
```diff
@@ -201,11 +208,158 @@ func NewConfigFactory(
+}
+
+func (c *ConfigFactory) invalidatePredicatesForPv(pv *v1.PersistentVolume) {
+	if c.enableEquivalenceClassCache {
```
You're already checking this in handler function - you can remove it from here.

resouer replied:
I just removed this check from handler functions, for consistency.

wojtek-t replied:
I think removing them from here would be better, because in case of disabled cache you're unnecessarily doing some work of casting etc.

resouer replied:
OK, that make sense

wojtek-t commented on plugin/pkg/scheduler/factory/factory.go:
```diff
@@ -201,11 +208,158 @@ func NewConfigFactory(
+}
+
+func (c *ConfigFactory) invalidatePredicatesForPvc(pvc *v1.PersistentVolumeClaim) {
+	if c.enableEquivalenceClassCache {
```
You're already checking this in handler function - you can remove it from here.

resouer replied:
Thanks. I just wondering is there anyway to enable this feature in integration & e2e tests? (I've manually enabled it in unit test)

wojtek-t commented on plugin/pkg/scheduler/factory/factory.go:
```diff
@@ -201,11 +208,158 @@ func NewConfigFactory(
+	}
+}
+
+func (c *ConfigFactory) onServiceUpdate(oldObj interface{}, newObj interface{}) {
```
Can you please sort functions in the order "add, update, delete" (for services?)

wojtek-t commented on plugin/pkg/scheduler/factory/factory.go:
```diff
@@ -201,11 +208,158 @@ func NewConfigFactory(
+		oldService := oldObj.(*v1.Service)
+		newService := newObj.(*v1.Service)
+		if !reflect.DeepEqual(oldService.Spec.Selector, newService.Spec.Selector) {
+			c.equivalencePodCache.InvalidateCachedPredicateItemOfAllNodes(sets.NewString("ServiceAffinity"))
```
Instead of creating the sets.String on every call of this function, you can simply define high level variable:
```
var serviceAffinitySet = sets.NewString("ServiceAffinity")
```
and just use it here.

In what you are doing we will waste a loooot of memory allocations and GC time.

wojtek-t commented on plugin/pkg/scheduler/factory/factory.go:
```diff
@@ -201,11 +208,158 @@ func NewConfigFactory(
+
+func (c *ConfigFactory) onServiceAdd(obj interface{}) {
+	if c.enableEquivalenceClassCache {
+		c.equivalencePodCache.InvalidateCachedPredicateItemOfAllNodes(sets.NewString("ServiceAffinity"))
```
Same here

wojtek-t commented on plugin/pkg/scheduler/factory/factory.go:
```diff
@@ -201,11 +208,158 @@ func NewConfigFactory(
+}
+
+func (c *ConfigFactory) onServiceDelete(obj interface{}) {
+	if c.enableEquivalenceClassCache {
```
and here and everywhere below

resouer replied:
Fixed!

wojtek-t commented on plugin/pkg/scheduler/factory/factory.go:
```diff
@@ -307,6 +510,63 @@ func (c *ConfigFactory) updateNodeInCache(oldObj, newObj interface{}) {
+func (c *ConfigFactory) invalidateCachedPredicatesOnNodeUpdate(newNode *v1.Node, oldNode *v1.Node) {
+	if c.enableEquivalenceClassCache {
+		// Begin to update equivalence cache based on node update
+		invalidPredicates := sets.NewString()
```
This seems to be the only place where you need it. But please add a TODO to lazily initialize it.

wojtek-t replied:
In most of cases (heartbeats) it will still be empty at the end.

wojtek-t commented on plugin/pkg/scheduler/factory/factory_test.go:
```diff
@@ -63,6 +63,8 @@ func TestCreate(t *testing.T) {
 		informerFactory.Apps().V1beta1().StatefulSets(),
 		informerFactory.Core().V1().Services(),
 		v1.DefaultHardPodAffinitySymmetricWeight,
+		// enable equivalence class cache
```
Can you make this a constant in this file, like:
```
const enableEquivalenceCache = false
```
and use it everywhere?

resouer replied:
Yes, but do you think it reasonable enable this in tests? (or, it's a little bit dangerous for 1.7)

wojtek-t replied:
I would say let's try unit & integration.

wojtek-t commented on test/integration/scheduler/scheduler_test.go:
```diff
@@ -256,6 +256,8 @@ func TestUnschedulableNodes(t *testing.T) {
 		informerFactory.Apps().V1beta1().StatefulSets(),
 		informerFactory.Core().V1().Services(),
 		v1.DefaultHardPodAffinitySymmetricWeight,
+		// enable equivalence class cache
```
Can you make this a constant (and use it both here and in extender test)?

resouer replied:
yes!

resouer commented:
@k8s-bot pull-kubernetes-unit test this

wojtek-t commented on plugin/pkg/scheduler/factory/factory.go:
```diff
@@ -257,6 +410,29 @@ func (c *ConfigFactory) updatePodInCache(oldObj, newObj interface{}) {
+				// MatchInterPodAffinity need to be reconsidered for this node,
+				// as well as all nodes in its same failure domain.
+				c.equivalencePodCache.InvalidateCachedPredicateItemOfAllNodes(
+					sets.NewString("MatchInterPodAffinity"))
```
and here

wojtek-t commented on plugin/pkg/scheduler/factory/factory.go:
```diff
@@ -257,6 +410,29 @@ func (c *ConfigFactory) updatePodInCache(oldObj, newObj interface{}) {
+			// if requested container resource changed, invalidate GeneralPredicates of this node
+			if !reflect.DeepEqual(predicates.GetResourceRequest(newPod),
+				predicates.GetResourceRequest(oldPod)) {
+				c.equivalencePodCache.InvalidateCachedPredicateItem(
```
Can you do the same you did with "serviceAffinitySet" ?

wojtek-t commented on plugin/pkg/scheduler/factory/factory.go:
```diff
@@ -278,6 +454,29 @@ func (c *ConfigFactory) deletePodFromCache(obj interface{}) {
+		// as well as all nodes in its same failure domain.
+		// TODO(resouer) can we just do this for nodes in the same failure domain
+		c.equivalencePodCache.InvalidateCachedPredicateItemOfAllNodes(
+			sets.NewString("MatchInterPodAffinity"))
```
Can you do the same you did with "serviceAffinitySet" ?

wojtek-t commented on plugin/pkg/scheduler/factory/factory.go:
```diff
@@ -278,6 +454,29 @@ func (c *ConfigFactory) deletePodFromCache(obj interface{}) {
+		for _, volume := range pod.Spec.Volumes {
+			if volume.GCEPersistentDisk != nil || volume.AWSElasticBlockStore != nil ||
+				volume.RBD != nil || volume.ISCSI != nil {
+				c.equivalencePodCache.InvalidateCachedPredicateItem(
```
Can you do the same you did with "serviceAffinitySet" ?

resouer replied:
Yes!

resouer commented:
@k8s-bot pull-kubernetes-e2e-gce-etcd3 test this
@k8s-bot pull-kubernetes-unit test this
@k8s-bot pull-kubernetes-federation-e2e-gce test this

resouer commented:
@wojtek-t Just fixed all comments and rebased.

resouer commented:
unit test flake ref #36238

sakeven commented on pkg/features/kube_features.go:
```diff
@@ -108,6 +108,12 @@ const (
+	// alpha: v1.7
+	//
+	// Enable equivalence class cache for scheduler.
+	EnableEquvalenceClassCache utilfeature.Feature = "EnableEquvalenceClassCache"
```
Is Equvalence misspelling in `EnableEquvalenceClassCache `?

resouer replied:
Seems so :/ fixing

wojtek-t commented:
@resouer - sorry for long delay but I was on 3 weeks vacation.

Can you please rebase the PR and I will do (hopefully the last) pass of review. I would really like to be merged soon.

resouer commented:
@wojtek-t Yes, rebase is on the way

wojtek-t commented:
@resouer - cool. Please ping me when it will be ready. I will treat it as a priority this time (I don't expect many more comments this time, since we already had few passes of review).

Commit:
Update generated bazel

resouer commented:
ping @wojtek-t Rebase nits were fixed, please review

wojtek-t commented on plugin/pkg/scheduler/factory/factory.go:
```diff
@@ -307,6 +515,64 @@ func (c *ConfigFactory) updateNodeInCache(oldObj, newObj interface{}) {
+		switch {
+		case !reflect.DeepEqual(oldNode.Status.Allocatable, newNode.Status.Allocatable):
+			invalidPredicates.Insert("GeneralPredicates") // "PodFitsResources"
+		case !reflect.DeepEqual(oldNode.GetLabels(), newNode.GetLabels()):
```
I think that we generally should check all the cases here (e.g.. if the previous were true, this one should also be handled).
This is not how switch works - it breaks on the first matched case, so I think using switch here is incorrect.

wojtek-t commented on plugin/pkg/scheduler/factory/factory.go:
```diff
@@ -307,6 +515,64 @@ func (c *ConfigFactory) updateNodeInCache(oldObj, newObj interface{}) {
+					}
+				}
+			}
+		case !reflect.DeepEqual(oldNode.GetName(), newNode.GetName()):
```
You can remove this - it's forbidden to change object name.

wojtek-t reviewed (commented):
Two more comments - once applied, I think this is good to go.

resouer commented:
@wojtek-t Nits fixed, please review.

wojtek-t commented:
LGTM.

Though you unfortunately need to rebase. Please ping me when you rebase it - I will do you minor sanity check and approve.

resouer commented:
/test pull-kubernetes-e2e-gce-etcd3
/test pull-kubernetes-kubemark-e2e-gce

resouer commented:
/test pull-kubernetes-kubemark-e2e-gce

gyliu513 commented:
/sig scheduling

resouer commented:
@wojtek-t rebase done, please do a final review

wojtek-t commented:
/lgtm
/approve no-issue
